sklearn/preprocessing/_encoders.py

```diff
@@ -89,55 +88,49 @@ def _fit(self, X, handle_unknown='error'):
         for i in range(n_features):
             Xi = X_list[i]
             if self._categories == 'auto':
-                cats = _encode(Xi)
+                cats = _nanencode(Xi, missing_values=missing_values)
             else:
                 cats = np.array(self._categories[i], dtype=Xi.dtype)
                 if Xi.dtype != object:
                     if not np.all(np.sort(cats) == cats):
                         raise ValueError("Unsorted categories are not "
                                          "supported for numerical categories")
                 if handle_unknown == 'error':
-                    diff = _encode_check_unknown(Xi, cats)
-                    if diff:
+                    try:
+                        _nanencode(Xi, cats, encode=True,
+                                   missing_values=missing_values)
+                    except ValueError as e:
+                        diff = e.args[0]
```

Comment:
I don't think we can assume that the ValueError being raised is the one you explicitly raise. ValueErrors are the most common unintended error in case of logical failure.